### PR TITLE
search: add topQueries to graphqlbackend

### LINKED ISSUE #3256: search: add endpoint returning top N recent queries
There should be an endpoint in graphqlbackend that returns something like this:
```
select query, count(*) from recent_searches group by query order by count desc limit 100;
```

### PR DESCRIPTION
**Pull request by ijt**
Fixes #3256.

This change also includes some refactoring of code working with the `recent_searches` table to stop using global variables since that led to a race condition in a test.

Test plan: Added a unit test for `TopQueries()`.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/top_queries.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	sort.Slice(uniques, func(i, j int) bool {
+		hi := histo[uniques[i]]
+		hj := histo[uniques[j]]
+		switch {
+		case hi > hj:
+			return true
+		case hi < hj:
+			return false
+		default:
+			return uniques[i] < uniques[j]
```
Comment (ijsnow) [inline cmd/frontend/graphqlbackend/top_queries.go]: Why the switch instead of a single if?

```go
if hi == hj {
  return uniques[i] < uniques[j]
}
return hi > hj
```

Comment (ijt) [reply cmd/frontend/graphqlbackend/top_queries.go]: It's needed to impose a deterministic order in cases where two or more queries have the same count.

File: cmd/frontend/db/recent_searches.go [hunk 2]
```diff
@@ -2,21 +2,21 @@ package db
...
 import (
 	"context"
+	"database/sql"
+
 	"github.com/pkg/errors"
-	"github.com/sourcegraph/sourcegraph/pkg/db/dbconn"
 )
 
-type recentSearches struct{}
+type RecentSearches struct {
```
Comment (beyang) [inline cmd/frontend/db/recent_searches.go]: It looks like this is not referenced anywhere in the new code. Is there a test we could add to ensure this is actually set? Or do we feel confident that it's obvious enough that this won't regress in the future?

Comment (ijt) [reply cmd/frontend/db/recent_searches.go]: This is a tough thing to test in an automated way because the code runs in init and involves globals. It would have to be some sort of e2e thing, probably not worth it.

Anyway, I think the risk is low here. Worst case, it regresses and we don't have queries showing up in the recent_searches table. So we notice that and turn it back on. No search results are affected.

Comment (ijt) [reply cmd/frontend/db/recent_searches.go]: Also I fixed it in 3b3af5f.

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 3]
```diff
@@ -50,7 +51,10 @@ func (prometheusTracer) TraceField(ctx context.Context, label, typeName, fieldNa
 
 func init() {
 	var err error
-	GraphQLSchema, err = graphql.ParseSchema(Schema, &schemaResolver{}, graphql.Tracer(prometheusTracer{}))
+	sr := &schemaResolver{
+		recentSearches: &db.RecentSearches{DB: dbconn.Global},
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/graphqlbackend.go]: There was a nil check for this global in the old code. What guarantee do we have that it's set here? Do we need a nil check here that will print a (potentially fatal) error here?

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 4]
```diff
@@ -138,7 +142,15 @@ func (r *nodeResolver) ToSite() (*siteResolver, bool) {
 	return n, ok
 }
 
-type schemaResolver struct{}
+type RecentSearchesTracker interface {
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/graphqlbackend.go]: Add a docstring for this interface and each method.

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 5]
```diff
@@ -138,7 +142,15 @@ func (r *nodeResolver) ToSite() (*siteResolver, bool) {
...
+	Add(ctx context.Context, q string) error
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/graphqlbackend.go]: What's `q` here? Is it "query"? Probably worth to be more descriptive in the interface declaration.

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 6]
```diff
@@ -138,7 +142,15 @@ func (r *nodeResolver) ToSite() (*siteResolver, bool) {
...
+	DeleteExcessRows(ctx context.Context, limit int) error
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/graphqlbackend.go]: Why is this necessary? When should it be invoked?

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 7]
```diff
@@ -138,7 +142,15 @@ func (r *nodeResolver) ToSite() (*siteResolver, bool) {
...
+	Get(ctx context.Context) ([]string, error)
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/graphqlbackend.go]: How many does this return? If it returns multiple, would `List` be a better name? Or even more explicit (e.g., `ListLast10`)?

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 8]
```diff
@@ -138,7 +142,15 @@ func (r *nodeResolver) ToSite() (*siteResolver, bool) {
...
+}
+
+type schemaResolver struct {
+	recentSearches RecentSearchesTracker
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/graphqlbackend.go]: Adding this here seems inconsistent with the other graphqlbackend resolvers (which I assume are globals). If this is a step toward dependency injection, that's fine, but I think it's worth calling out in a comment on `schemaResolver` what the current state is. E.g., "schemaResolver references subresolvers, some of which are globals and some of which are fields in the struct. Eventually, they should all be dependency-injected, but in the meantime, ... etc".

File: cmd/frontend/graphqlbackend/top_queries.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import (
+	"context"
+	"sort"
+
+	"github.com/pkg/errors"
+)
+
+// TopQueries returns the top most frequent recent queries.
+func (s *schemaResolver) TopQueries(ctx context.Context, args *struct{ Limit int32 }) ([]queryCountResolver, error) {
+	searches, err := s.recentSearches.Get(ctx)
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/top_queries.go]: Why not forward the `Limit` to `s.recentSearches.Get`? What's default limit we expect to get back from `s.recentSearches.Get`?

File: cmd/frontend/graphqlbackend/schema.go [hunk 10]
```diff
@@ -791,6 +796,15 @@ type Query {
     dotcom: DotcomQuery!
 }
 
+# A query and an associated number of times it occurred.
+type QueryCount {
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/schema.go]: Is this used anywhere?

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 11]
```diff
@@ -798,6 +803,15 @@ type Query {
     dotcom: DotcomQuery!
 }
 
+# A query and an associated number of times it occurred.
+type QueryCount {
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/schema.graphql]: Is this used anywhere?

File: cmd/frontend/graphqlbackend/search.go [hunk 12]
```diff
@@ -80,14 +83,16 @@ func (r *schemaResolver) Search(args *struct {
 	}, nil
 }
 
-func addQueryToSearchesTable(q string) {
+func addQueryToSearchesTable(rs RecentSearchesTracker, q string) {
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/search.go]: It seems like this would make more sense as either a method on `schemaResolver` or a method on `RecentSearchesTracker`, rather than a standalone function now.

File: cmd/frontend/graphqlbackend/top_queries.go [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "getting recent searches from database")
+	}
+	histo := make(map[string]int32)
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/top_queries.go]: Why not do the "unique + limit" logic in the DB query, rather than here in application logic?

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 3 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: Good catch. This happens before the call to `dbconn.ConnectToDB` in `serve_cmd.go`, so `dbconn.Global` is nil here (verified it).

As a short-term solution, I changed the db field to a func so it gets the current value of the global db conn. I manually verified that my search queries are showing up in the `recent_searches` table.

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 4 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: Done.

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 7 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: It returns all of them. I renamed it to List.

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 6 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: Nick asked about how large the table would get, so I added this as a way to keep the table from growing to an unbounded size.

It should be invoked often enough to prevent the `recent_searches` table from eating up all the disk space. Currently it is invoked with every call to `Add` (renamed to `Log`), but it could be done less frequently.

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 8 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: Yes, it's inconsistent, and it is a step towards dependency injection.

I added a comment along the lines you suggested.

File: cmd/frontend/graphqlbackend/top_queries.go (hunk 9 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/top_queries.go]: The limit for `TopQueries` is for the number of unique query strings. That doesn't make sense to forward to `s.recentSearches.Get` since `s.recentSearches.Get` returns all the query strings including redundant ones.

`s.recentSearches.Get` has no default limit.

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 11 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/schema.graphql]: Yes, it's used in the return for `topQueries`.

File: cmd/frontend/graphqlbackend/search.go (hunk 12 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/search.go]: Done.

File: cmd/frontend/graphqlbackend/top_queries.go (hunk 13 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/top_queries.go]: Done.

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 5 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: Done.

File: cmd/frontend/graphqlbackend/search.go [hunk 14]
```diff
@@ -69,7 +69,10 @@ func (r *schemaResolver) Search(args *struct {
 		return newSearcherResolver(query2)
 	}
 
-	go addQueryToSearchesTable(args.Query)
+	// TODO(ijt): remove this goroutine leak. One simple approach would be to issue a separate, concurrent query
+	// from the typescript frontend to a new graphql mutation that adds a query to the recent_searches table.
+	go r.addQueryToSearchesTable(args.Query)
```
Comment (ijsnow) [inline cmd/frontend/graphqlbackend/search.go]: What makes this a leak? It should start and then close itself when it's done, right? I might not know enough about goroutines to see the problem.

Comment (ijsnow) [reply cmd/frontend/graphqlbackend/search.go]: Should this be above the `if nested { return ... }` so that it gets ran no matter what?

**Comment by ijt** [comment]
@beyang, changes are up. PTAL.

File: cmd/frontend/graphqlbackend/top_queries.go (hunk 13 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/top_queries.go]: Done.

File: cmd/frontend/graphqlbackend/search.go (hunk 14 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/search.go]: It's a potential leak. One way it could happen is if the `Cleanup` calls stall. Then these goroutines would pile up. It could also mean that some of the row insertions time out and some new queries don't make it into the db.

I think a better way would be to have a cleanup goroutine with lifespan equal to that of `schemaResolver`. This goroutine would wake up maybe once every hour and remove rows until the physical size of the `recent_searches` table is under some threshold such as 10M.

Comment (ijt) [reply cmd/frontend/graphqlbackend/search.go]: > Should this be above the `if nested { return ... }` so that it gets ran no matter what?

Good call, but that's beyond the scope of this change.

**Comment by ijt** [comment]
Hold on, I have to fix the ordering...

**Comment by ijt** [comment]
Fixed the ordering. PTAL.

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 15]
```diff
@@ -138,7 +141,31 @@ func (r *nodeResolver) ToSite() (*siteResolver, bool) {
...
+	Top(ctx context.Context, n int32) ([]string, []int32, error)
+
+	// Cleanup removes old entries such that there are no more than limit remaining.
+	Cleanup(ctx context.Context, limit int) error
+}
+
+// schemaResolver handles all GraphQL queries for Sourcegraph.  To do this, it
+// uses subresolvers, some of which are globals and some of which are fields on
+// schemaResolver. Eventually, they should all be fields (i.e., dependency
+// injected), but that is being done gradually.
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/graphqlbackend.go]:
```suggestion
// injected), but that is being done gradually. Currently, only `recentSearches` is dependency-injected.
```

File: cmd/frontend/db/recent_searches.go [hunk 16]
```diff
@@ -67,3 +63,24 @@ func (*recentSearches) Get(ctx context.Context) ([]string, error) {
 	}
 	return qs, nil
 }
+
+// Top returns the top n queries in the recent_searches table.
+func (rs *RecentSearches) Top(ctx context.Context, n int32) ([]string, []int32, error) {
+	sel := `SELECT query, COUNT(*) FROM recent_searches GROUP BY query ORDER BY 2 DESC, query ASC LIMIT $1`
```
Comment (beyang) [inline cmd/frontend/db/recent_searches.go]: Can the `ORDER BY` clause be a named column (so this is robust to column ordering in the future)?

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 17]
```diff
@@ -138,7 +141,31 @@ func (r *nodeResolver) ToSite() (*siteResolver, bool) {
 	return n, ok
 }
 
-type schemaResolver struct{}
+// StringLogger describes something that can log strings, list them and also
+// clean up to make sure they don't use too much storage space.
+type StringLogger interface {
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/graphqlbackend.go]: Is this interface really necessary? It is only used in one place. Why not just use the concrete type?

Comment (beyang) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: Oh, I see; you mock it in a test. The name should be more specific to its purpose. E.g., `RecentSearches` or `RecentSearchesSvc`. `StringLogger` seems overly generic, since we don't need just any string logger, but one specifically representing the recent searches backend.

File: cmd/frontend/graphqlbackend/search_test.go [hunk 18]
```diff
@@ -99,6 +108,15 @@ func TestSearch(t *testing.T) {
 	}
 }
 
+type RecentSearchesIgnorer struct{}
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/search_test.go]:
```suggestion
type NoOpRecentSearches struct{}
```

Comment (beyang) [reply cmd/frontend/graphqlbackend/search_test.go]: "Nop" seems more conventional in Go (e.g., https://golang.org/pkg/io/ioutil/#NopCloser)

File: cmd/frontend/db/recent_searches.go [hunk 19]
```diff
@@ -2,21 +2,20 @@ package db
...
 
-type recentSearches struct{}
+type RecentSearches struct{}
 
-// Add inserts the query q to the recentSearches table in the db.
-func (*recentSearches) Add(ctx context.Context, q string) error {
+// Log inserts the query q into to the recent_searches table in the db.
+func (rs *RecentSearches) Log(ctx context.Context, q string) error {
 	insert := `INSERT INTO recent_searches (query) VALUES ($1)`
-	if dbconn.Global == nil {
```
Comment (beyang) [inline cmd/frontend/db/recent_searches.go]: Do we need to preserve the nil check given that we're still using `dbconn.Global`?

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 17 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: We don't need it to update that table when we are running tests that don't check that table, so I think the more generic name makes sense here.

File: cmd/frontend/graphqlbackend/search_test.go (hunk 18 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/search_test.go]: Done.

File: cmd/frontend/db/recent_searches.go (hunk 19 above)
Comment (ijt) [reply cmd/frontend/db/recent_searches.go]: I don't think so. `dbconn.Global` only defaults to nil during unit tests, and for those I inject mocks for `schemaResolver.recentSearches` when it's needed.

File: cmd/frontend/db/recent_searches.go (hunk 16 above)
Comment (ijt) [reply cmd/frontend/db/recent_searches.go]: Done.

File: cmd/frontend/db/recent_searches.go (hunk 19 above)
Comment (beyang) [reply cmd/frontend/db/recent_searches.go]: It looks like initialization of that variable takes place in this function: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@cc5de96d4b3e20569d1791992404303943fbd5b0/-/blob/pkg/db/dbconn/dbconn.go#L46:6&tab=references. And it is unclear to me if there's any guarantee on when that function is called relative to when this method might be invoked. The need for a nil check for `dbconn.Global` here is code smell, but I think it's needed given that we're not really sure if it is nil or non-nil at this point.

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 17 above)
Comment (beyang) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: The name "recent searches" refers not to the underlying DB table but to the API scope. For example, what if we want to add an additional API endpoints related to recent searches (e.g., some aggregate statistic about recent searches or list users who have completed recent searches). Would those methods go into separate service structs or would they be added here?

File: cmd/frontend/db/recent_searches.go (hunk 19 above)
Comment (ijt) [reply cmd/frontend/db/recent_searches.go]: I'm not sure I understand how this is different from other places nearby where dbconn.Global is used without a nil check. For example https://sourcegraph.com/github.com/sourcegraph/sourcegraph@cc5de96/-/blob/cmd/frontend/db/repos.go#L97. I can put these nil checks back, but then shouldn't we add them all over the place in the rest of `frontend/db/...`?

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 17 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: Those methods could go into `frontent/db.RecentSearches` and it would still conform to this interface. I think this interface makes sense here with the name it currently has.
